File: ios/CCOAN/Models/Project.swift
import Foundation

enum ProjectStatus: String, Codable, CaseIterable {
    case backlog
    case todo
    case in_progress
    case done
    case archived

    var displayName: String {
        switch self {
        case .backlog: return "Backlog"
        case .todo: return "To Do"
        case .in_progress: return "In Progress"
        case .done: return "Done"
        case .archived: return "Archived"
        }
    }
}

enum Priority: String, Codable, CaseIterable {
    case low
    case medium
    case high
    case urgent

    var displayName: String {
        rawValue.capitalized
    }

    var color: String {
        switch self {
        case .low: return "#22c55e"
        case .medium: return "#eab308"
        case .high: return "#f97316"
        case .urgent: return "#ef4444"
        }
    }
}

enum LabelColor: String, Codable, CaseIterable {
    case red, orange, yellow, green, blue, purple, pink, gray

    var hexColor: String {
        switch self {
        case .red: return "#ef4444"
        case .orange: return "#f97316"
        case .yellow: return "#eab308"
        case .green: return "#22c55e"
        case .blue: return "#3b82f6"
        case .purple: return "#a855f7"
        case .pink: return "#ec4899"
        case .gray: return "#6b7280"
        }
    }
}

struct Label: Codable, Identifiable {
    let id: String
    let name: String
    let color: LabelColor
}

struct Subtask: Codable, Identifiable {
    let id: String
    let title: String
    var completed: Bool
}

struct Comment: Codable, Identifiable {
    let id: String
    let userId: String
    let userName: String
    let text: String
    let createdAt: String
}

struct Attachment: Codable, Identifiable {
    let id: String
    let name: String
    let url: String
    let type: String
    let size: Int
    let uploadedAt: String
    let uploadedBy: String
    let uploadedByName: String?
}

struct ActivityLogEntry: Codable, Identifiable {
    let id: String
    let userId: String
    let userName: String
    let action: String
    let details: String?
    let timestamp: String
}

struct RecurrenceConfig: Codable {
    let type: RecurrenceType
    let endDate: String?
    let lastGenerated: String?
}

enum RecurrenceType: String, Codable, CaseIterable {
    case none
    case daily
    case weekly
    case biweekly
    case monthly
}

struct Project: Codable, Identifiable {
    let id: String
    let departmentId: String
    var departmentName: String?
    var title: String
    var description: String
    var status: ProjectStatus
    var priority: Priority
    var assigneeId: String?
    var assigneeName: String?
    var assigneeIds: [String]?
    var assigneeNames: [String]?
    let createdBy: String
    var creatorName: String?
    var dueDate: String?
    var labels: [Label]
    var subtasks: [Subtask]
    var attachments: [Attachment]
    var comments: [Comment]
    var activityLog: [ActivityLogEntry]
    var estimatedHours: Double?
    var loggedHours: Double?
    var blockedBy: [String]?
    var order: Int
    var recurrence: RecurrenceConfig?
    let metadata: MetadataTimestamps?

    enum CodingKeys: String, CodingKey {
        case id, departmentId, departmentName, title, description, status, priority
        case assigneeId, assigneeName, assigneeIds, assigneeNames
        case createdBy, creatorName, dueDate, labels, subtasks, attachments
        case comments, activityLog, estimatedHours, loggedHours, blockedBy
        case order, recurrence, metadata
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try container.decode(String.self, forKey: .id)
        self.departmentId = try container.decode(String.self, forKey: .departmentId)
        self.departmentName = try container.decodeIfPresent(String.self, forKey: .departmentName)
        self.title = try container.decode(String.self, forKey: .title)
        self.description = try container.decodeIfPresent(String.self, forKey: .description) ?? ""
        self.status = try container.decode(ProjectStatus.self, forKey: .status)
        self.priority = try container.decodeIfPresent(Priority.self, forKey: .priority) ?? .medium
        self.assigneeId = try container.decodeIfPresent(String.self, forKey: .assigneeId)
        self.assigneeName = try container.decodeIfPresent(String.self, forKey: .assigneeName)
        self.assigneeIds = try container.decodeIfPresent([String].self, forKey: .assigneeIds)
        self.assigneeNames = try container.decodeIfPresent([String].self, forKey: .assigneeNames)
        self.createdBy = try container.decode(String.self, forKey: .createdBy)
        self.creatorName = try container.decodeIfPresent(String.self, forKey: .creatorName)
        self.dueDate = try container.decodeIfPresent(String.self, forKey: .dueDate)
        self.labels = try container.decodeIfPresent([Label].self, forKey: .labels) ?? []
        self.subtasks = try container.decodeIfPresent([Subtask].self, forKey: .subtasks) ?? []
        self.attachments = try container.decodeIfPresent([Attachment].self, forKey: .attachments) ?? []
        self.comments = try container.decodeIfPresent([Comment].self, forKey: .comments) ?? []
        self.activityLog = try container.decodeIfPresent([ActivityLogEntry].self, forKey: .activityLog) ?? []
        self.estimatedHours = try container.decodeIfPresent(Double.self, forKey: .estimatedHours)
        self.loggedHours = try container.decodeIfPresent(Double.self, forKey: .loggedHours)
        self.blockedBy = try container.decodeIfPresent([String].self, forKey: .blockedBy)
        self.order = try container.decodeIfPresent(Int.self, forKey: .order) ?? 0
        self.recurrence = try container.decodeIfPresent(RecurrenceConfig.self, forKey: .recurrence)
        self.metadata = try container.decodeIfPresent(MetadataTimestamps.self, forKey: .metadata)
    }
}

struct CreateProjectRequest: Encodable {
    let departmentId: String
    let title: String
    let description: String?
    let status: String
    let priority: String
    let assigneeId: String?
    let assigneeIds: [String]?
    let dueDate: String?
    let labels: [Label]?
    let subtasks: [SubtaskInput]?
    let estimatedHours: Double?
    let recurrence: RecurrenceInput?
}

struct SubtaskInput: Encodable {
    let title: String
}

struct RecurrenceInput: Encodable {
    let type: String
    let endDate: String?
}

struct UpdateProjectRequest: Encodable {
    let id: String
    let title: String?
    let description: String?
    let status: String?
    let priority: String?
    let assigneeId: String?
    let assigneeIds: [String]?
    let dueDate: String?
    let labels: [Label]?
    let subtasks: [Subtask]?
    let estimatedHours: Double?
    let loggedHours: Double?
    let blockedBy: [String]?
}

struct ReorderProjectRequest: Encodable {
    let projectId: String
    let newStatus: String
    let newOrder: Int
}

struct AddCommentRequest: Encodable {
    let text: String
}

File: android/app/src/main/java/com/ccoan/ny/data/models/Project.kt
package com.ccoan.ny.data.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

enum class ProjectStatus {
    @Json(name = "backlog") BACKLOG,
    @Json(name = "todo") TODO,
    @Json(name = "in_progress") IN_PROGRESS,
    @Json(name = "done") DONE,
    @Json(name = "archived") ARCHIVED;

    val displayName: String
        get() = when (this) {
            BACKLOG -> "Backlog"
            TODO -> "To Do"
            IN_PROGRESS -> "In Progress"
            DONE -> "Done"
            ARCHIVED -> "Archived"
        }

    companion object {
        fun fromString(value: String): ProjectStatus {
            return entries.find { it.name.equals(value, ignoreCase = true) } ?: BACKLOG
        }
    }
}

enum class Priority {
    @Json(name = "low") LOW,
    @Json(name = "medium") MEDIUM,
    @Json(name = "high") HIGH,
    @Json(name = "urgent") URGENT;

    val displayName: String
        get() = name.lowercase().replaceFirstChar { it.uppercase() }

    val colorHex: String
        get() = when (this) {
            LOW -> "#22C55E"
            MEDIUM -> "#EAB308"
            HIGH -> "#F97316"
            URGENT -> "#EF4444"
        }
}

enum class LabelColor {
    @Json(name = "red") RED,
    @Json(name = "orange") ORANGE,
    @Json(name = "yellow") YELLOW,
    @Json(name = "green") GREEN,
    @Json(name = "blue") BLUE,
    @Json(name = "purple") PURPLE,
    @Json(name = "pink") PINK,
    @Json(name = "gray") GRAY;

    val hexColor: String
        get() = when (this) {
            RED -> "#EF4444"
            ORANGE -> "#F97316"
            YELLOW -> "#EAB308"
            GREEN -> "#22C55E"
            BLUE -> "#3B82F6"
            PURPLE -> "#A855F7"
            PINK -> "#EC4899"
            GRAY -> "#6B7280"
        }
}

@JsonClass(generateAdapter = true)
data class Label(
    val id: String,
    val name: String,
    val color: LabelColor
)

@JsonClass(generateAdapter = true)
data class Subtask(
    val id: String,
    val title: String,
    var completed: Boolean
)

@JsonClass(generateAdapter = true)
data class Comment(
    val id: String,
    val userId: String,
    val userName: String,
    val text: String,
    val createdAt: String
)

@JsonClass(generateAdapter = true)
data class Attachment(
    val id: String,
    val name: String,
    val url: String,
    val type: String,
    val size: Int,
    val uploadedAt: String,
    val uploadedBy: String,
    val uploadedByName: String? = null
)

@JsonClass(generateAdapter = true)
data class ActivityLogEntry(
    val id: String,
    val userId: String,
    val userName: String,
    val action: String,
    val details: String? = null,
    val timestamp: String
)

enum class RecurrenceType {
    @Json(name = "none") NONE,
    @Json(name = "daily") DAILY,
    @Json(name = "weekly") WEEKLY,
    @Json(name = "biweekly") BIWEEKLY,
    @Json(name = "monthly") MONTHLY
}

@JsonClass(generateAdapter = true)
data class RecurrenceConfig(
    val type: RecurrenceType,
    val endDate: String? = null,
    val lastGenerated: String? = null
)

@JsonClass(generateAdapter = true)
data class Project(
    val id: String,
    val departmentId: String,
    val departmentName: String? = null,
    var title: String,
    var description: String = "",
    var status: ProjectStatus = ProjectStatus.BACKLOG,
    var priority: Priority = Priority.MEDIUM,
    var assigneeId: String? = null,
    var assigneeName: String? = null,
    var assigneeIds: List<String>? = null,
    var assigneeNames: List<String>? = null,
    val createdBy: String,
    var creatorName: String? = null,
    var dueDate: String? = null,
    var labels: List<Label> = emptyList(),
    var subtasks: List<Subtask> = emptyList(),
    var attachments: List<Attachment> = emptyList(),
    var comments: List<Comment> = emptyList(),
    var activityLog: List<ActivityLogEntry> = emptyList(),
    var estimatedHours: Double? = null,
    var loggedHours: Double? = null,
    var blockedBy: List<String>? = null,
    var order: Int = 0,
    var recurrence: RecurrenceConfig? = null,
    val metadata: MetadataTimestamps? = null
)

@JsonClass(generateAdapter = true)
data class CreateProjectRequest(
    val departmentId: String,
    val title: String,
    val description: String? = null,
    val status: String = "backlog",
    val priority: String = "medium",
    val assigneeId: String? = null,
    val assigneeIds: List<String>? = null,
    val dueDate: String? = null,
    val labels: List<Label>? = null,
    val subtasks: List<SubtaskInput>? = null,
    val estimatedHours: Double? = null,
    val recurrence: RecurrenceInput? = null
)

@JsonClass(generateAdapter = true)
data class SubtaskInput(
    val title: String
)

@JsonClass(generateAdapter = true)
data class RecurrenceInput(
    val type: String,
    val endDate: String? = null
)

@JsonClass(generateAdapter = true)
data class UpdateProjectRequest(
    val id: String,
    val title: String? = null,
    val description: String? = null,
    val status: String? = null,
    val priority: String? = null,
    val assigneeId: String? = null,
    val assigneeIds: List<String>? = null,
    val dueDate: String? = null,
    val labels: List<Label>? = null,
    val subtasks: List<Subtask>? = null,
    val estimatedHours: Double? = null,
    val loggedHours: Double? = null,
    val blockedBy: List<String>? = null
)

@JsonClass(generateAdapter = true)
data class ReorderProjectRequest(
    val projectId: String,
    val newStatus: String,
    val newOrder: Int
)

@JsonClass(generateAdapter = true)
data class AddCommentRequest(
    val text: String
)
